[Wei Kit Khoong] iP

khoongwk opened the pull request.

Commit:
Add Gradle support

khoongwk committed:
Completed Increment Level-1 for greeting and echoing of commands. Option to exit using the command 'bye'.

khoongwk committed:
Completed increment level 2 with ability to store entries temporarily and list them out with the 'list' command. Any other input besides 'bye' would be treated as an entry to store.

khoongwk committed:
Completed increment level 3 with ability to mark a task as done.

khoongwk committed:
Completed Level 4 with tasks event, deadline, and todo. More input checking to verify command is valid.

khoongwk committed:
Completed level 5: Changed Input.txt and expected.txt

khoongwk committed:
Level 5: Incorporated DukeException for error handling.

khoongwk committed:
Completed Level 6: Delete Function

khoongwk committed:
Add Enums for Commands

If-Else Control flow and the substring() is used to determine the user's
input i.e 'bye', 'done', 'todo', etc.

However, this approach is messy and is not very readable.

Let's use Enums and switch-case statements to improve the code's readability.

khoongwk committed:
Add Save feature for Level 7

Tasks which are stored in the list while the program is run
is not saved and stored when the user exits the program.

Enabling a save function would improve the usability of this
program as tasks which were entered previously can be
remembered.

Let's providing a saving function which stores the Task list into
a text file when the user exits and retrieves this list when the
program is started up again.

khoongwk committed:
Fix issue with text file saving.

Working directory is different when building and running Duke in the IntelliJ IDE
compared to when running runtest.bat.

Modifications have to be made to the filepath specified depending on which
is the working directory.

Let's use a conditional expression to set the correct file path depending on what
was the original working directory.

khoongwk committed:
Add Date Time Formatting for user input

Date and Time of events and deadlines can take any format.

If Duke can understand certain formats of date and time input,
it can parse the String input to make it look more readable and
aesthetically pleasing.

To improve the user experience, let's enable Duke to process
specific time and date formats such as "dd/MM/yyyy HHmm" and
format it into a more pleasing look, such as "JULY 4 2021".

khoongwk committed:
Edit Text files for text-ui-test and other minor edits

Expected text file has changed due to new Save feature.

Replacing expected text would lead to a match between actual and expected text.

Let's make minor edits to text files as well as the main Duke code to clean up.

khoongwk committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	text-ui-test/input.txt

Conflict between master and branch-Level-8 input.txt.

Master input.txt contained a different event commmand
from branch-Level-8.

Let's choose branch-Level-8's event command as it
reflects the latest date and time capabilities of Duke.

khoongwk committed:
Add More Classes for OOP

Currently, the logic for Duke is all within the Duke class.

However, this design is bloated and should be compartmentalized
further using more OOP.

Let's introduce new classes such as Storage, TaskList, and Ui to
improve the structure of Duke and decentralize the logic to each
seperate component.

khoongwk committed:
Fix OOP Extensions

More Classes have been added to decentralise the logic,
such as the Storage, Parser, and Ui Class.

However, the classes were not feature complete, especially
the Ui Class.

Let's complete these classes, revamp the logic in the main
Duke class to use these classes, and test it out with text-ui-test.

khoongwk committed:
Package Classes into Duke and Task Packages

The numerous classes in the java directory has become
messy as the number of classes increased.

To improve the organization of the code base, packaging ought
to be done.

Let's package the classes into two main packages - task for the
Task Classes and duke for all the other classes. More packages
will be created in the future if complexity increases. Access modifiers
have been made package-private in the duke class for tighter
security.

Furthermore, minor edits to the code have been made such as
improved exception handling, and fully delegating console output
to the Ui class.

khoongwk committed:
Add JUnit Tests

Duke currently has no unit tests and relies on text-ui-test
to verify that the program works fine.

Adding unit tests using JUnit would improve overall code
quality by ensuring that methods in the different Duke
classes return the same output or throw the same exception.

Let's introduce a few JUnit test classes to test the funcionality
of the individual classes. More test classes could be included,
but the tests for Parser, Storage, and Task should be
sufficient for this current stage.

khoongwk committed:
Edit Ui Introduction Text

Ui shows commands available like event, todo,
deadline, list.

However, it missed other commands such as
delete, done, and bye.

Let's add the missing commands into the
introductory message.

khoongwk committed:
Add Find Command (Level 9)

Find Command skeleton has been part of Duke so far
but the logic has not been incorporated and tested.

Find command is useful for users to search any tasks
by their keyword(s).

To improve the usability of this program, let's provide
the logic of the find command, including testing via
text-ui-test.

khoongwk committed:
Edit code to match coding standard

Methods, variables have been aptly named according
to se-edu Java conventions.

However, there are sections where noise can be
eliminated by removing 'this' where it is unnecessary.

Let's refine the code base by removing the 'this' keyword
where it is unnecessary and make other edits where\
appropriate to meet the coding standard.

khoongwk committed:
Add Javadoc for all classes

Documentation of Classes and their methods
is important for projects where other developers
have to read code from other authors.

However, there is currently a lack of javadoc
documentation in this Duke project.

Thus, let's add javadoc documentation to all
classes and non-private methods where possible.

khoongwk committed:
Merge branch 'branch-Level-9'

khoongwk committed:
Merge branch 'branch-A-CodingStandard'

khoongwk committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/TaskList.java

jianhandev commented on src/main/java/duke/DateTimeProcessor.java:
```diff
@@ -0,0 +1,85 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.Month;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Optional;
```
Consider separating imports into groups, and in lexicological order.

khoongwk replied:
Doing either will fail gradle's checkstyle. 😢

jianhandev commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,165 @@
+                String input = ui.readCommand();
+                Parser.Command command = Parser.parseCommand(input);
+                switch (command) {
+                case BYE:
```
The case line should be at the same indentation level at the switch statement.

jianhandev commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Mark this task as done.
```
"Mark" should be "Marks"

jianhandev commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,114 @@
+        try {
+            File taskListFile = new File(filePath);
+            // Clear the pre-existing file if there is one.
+            if (taskListFile.exists()) {
+                taskListFile.delete();
+                taskListFile.createNewFile();
+            } else {
+                throw new IOException("No file found at the specified path.");
+            }
+            if (taskList.getSize() > 0) {
+                FileWriter taskWriter = new FileWriter(taskListFile);
+                for (Task t : taskList.getTaskList()) {
+                    taskWriter.write(t.getDescriptionForDatabase());
+                    taskWriter.write("\n");
+                }
+                taskWriter.close();
+            }
```
Consider separating different logical units with a line break.

khoongwk replied:
Thanks for the comments Jian Han 👍

khoongwk committed:
Merge remote-tracking branch 'origin/add-gradle-support'

khoongwk committed:
Add Gradle Support and Checkstyle

Gradle was not activated as well as checkstyle.

To automate future builds and improve the code quality,
let's enable Gradle and the checkstyle plugin.

khoongwk committed:
Add JavaFX Graphical User Interface

Duke has been using Console input to function thus far.

To improve the user experience, a graphical user interface
would appear as more aesthetic and appealing to users.

Thus, let's introduce a graphical user interface using the
JavaFX library. The GUI is launched from the Launcher
class, while the original console application is preserved
as the main method in Duke.java.

khoongwk committed:
Merge branch 'branch-Level-10'

khoongwk committed:
Add Assertions

Add assert statements in Parser and Storage
classes to improve code quality.

khoongwk committed:
Improve Code Quality

Duke's logic has been centred in the Duke class thus far.

However, the code quality is not very good as the getResponse
method is very long and does too many things.

In order to improve the quality of the code, let's abstract the logic
into 2 areas: 1) Parser class will process the user's input and
return the right Command. 2) Introduce Command classes which
can be executed to return a String response. Thus, the logic in Duke
would be abstracted out into Commands which are executed until
the command signals duke to stop running with the continueDuke()
method.

khoongwk committed:
Fix Minor issues in Duke

Tweaked file path system in Storage class.

khoongwk committed:
Merge pull request #1 from khoongwk/branch-A-Assertions

Branch A-Assertions

khoongwk committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	build.gradle
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/resources/images/DaDoge.jpg
#	src/main/resources/images/DaMusk.jpg

Resolved conflicts.

khoongwk committed:
Add BCD Extension C-Help, Fix Bugs

Users might be lost as to how to use Duke as there
is no guide provided to them.

Let's introduce a help command, which produces
useful instructions on what each command does
and how to use it. A help message is also produced
at the start of Duke.

Bugs were also fixed: Exception is now handled
when non-numbers are input for the delete and
done commands.

khoongwk committed:
Improve GUI

Components in JavaFX do not resize dynamically
when the main window is resized by the user.

To improve the look of the UI, let's tweak the fxml
files such that the scroll pane stretches to the width
and height of the parent main window.

khoongwk committed:
Improve GUI #2

Previous commit did not stage changes to GUI
correctly.

Let's stage all the changes to the GUI which will
improve the user experiences slightly.

khoongwk committed:
Add Ui Photo

Added screenshot of Ui of Duke and
reduced size of user image slightly to
improve readability of commands.

khoongwk committed:
Rework Storage Class

Storage has depended on Duke supplying a
correct file path of where the saved tasks were
stored.

However, Duke should not need to know the file
path and this function should be delegated to the
Storage class.

To improve the abstraction of Storage's function,
let's shift the file path to within the Storage class
such that everything related to the file I/O is
contained within the Storage class.

khoongwk committed:
Merge branch 'branch-ReworkStorage'

khoongwk committed:
Refactor Packages into One Main Package

Main folder had numerous packages but is suspected to cause
issues with the building of JAR file.

Let's refactor the packags such that all packages are within
a main 'duke' package. This would make the project structure
more organised.

khoongwk committed:
Enhance Find Command

Find command was finding exact matches to
keywords provided by the user.

To improve the usability of the find command,
let's make the searching of matching tasks
case-insensitive.

khoongwk committed:
Fix JAR file Creation

Duke works in Intellij but FXMLLoader.load() fails
when Duke is run as a JAR file.

File directory 'view' was capitalized by accident.
Let's fix this by renaming 'View' to 'view' under
src/main/resources.

khoongwk committed:
Fix JAR file Creation

Duke works in Intellij but FXMLLoader.load() fails
when Duke is run as a JAR file.

File directory 'view' was capitalized by accident.
Let's fix this by renaming 'View' to 'view' under
src/main/resources.

khoongwk committed:
Add User Guide

Currently there is no user guide available to users. Let's
add a user guide so that users can have a more
convenient experience when using Elon for the first time.

khoongwk committed:
Edit Readme

Github Readme is using the original template from Duke.

Let's edit it to suit the current state of Duke - in its renamed
form - Elon.

khoongwk committed:
Edit Order of Import

Minor edit to lexicographical order of imports
in EventCommand such that the project passes
checkstyle.

khoongwk committed:
Edit Ui.png file extension

Ui.PNG was not being detected by the CS2103T script.
Let's rename Ui.PNG to lowercase - Ui.png so that the
image would be detected and displayed under the iP
show case.

khoongwk committed:
Improve User Guide

User guide does not have anchor points to sections and expected outcomes
when commands are called.

To improve the navigability of and usefulness of the user guide, let's add in
anchor points to each section and expected outcomes when commands are
called such that users know what to expect while using Elon

khoongwk committed:
Fix Date Time Formatting

DateTimeFormatter Class represents minutes 0 - 9 in
single digits when it should be double digits i.e "12:09AM"
rather than "12:9 AM".

Let's fix the date time formatting such that zeroes are added
into the minutes representation where it is required.

khoongwk committed:
Edit User Guide

Minor formatting issue with Ui image in the user guide.

khoongwk committed:
Edit README

Line breaks were not working in the user guide and project README.

Let's add line breaks using backslashes.
